packages/jspsych/src/modules/plugin-api/KeyboardListenerAPI.ts

```diff
@@ -1,176 +1,149 @@
-          }
+    const listener: KeyboardListener = {
+      type: "keydown",
+      fn: (e: KeyboardEvent) => {
+        const rt =
+          (rt_method == "performance" ? performance.now() : audio_context.currentTime * 1000) -
+          startTime;
+        if (rt < minimum_valid_rt) {
+          return;
         }
-      }
 
-      // check if key was already held down
-      if (
-        (typeof parameters.allow_held_key === "undefined" || !parameters.allow_held_key) &&
-        valid_response
-      ) {
-        if (typeof this.held_keys[e.key] !== "undefined" && this.held_keys[e.key] == true) {
-          valid_response = false;
-        }
-        if (
-          !case_sensitive &&
-          typeof this.held_keys[e.key.toLowerCase()] !== "undefined" &&
-          this.held_keys[e.key.toLowerCase()] == true
-        ) {
-          valid_response = false;
-        }
-      }
+        const key = this.toLowerCaseIfInsensitive(e.key);
 
-      if (valid_response) {
-        // if this is a valid response, then we don't want the key event to trigger other actions
-        // like scrolling via the spacebar.
-        e.preventDefault();
-        var key = e.key;
-        if (!case_sensitive) {
-          key = key.toLowerCase();
-        }
-        parameters.callback_function({
-          key: key,
-          rt: rt_ms,
-        });
+        if (this.isResponseValid(valid_responses, allow_held_key, key)) {
+          // if this is a valid response, then we don't want the key event to trigger other actions
+          // like scrolling via the spacebar.
+          e.preventDefault();
 
-        if (this.keyboard_listeners.includes(listener_id)) {
-          if (!parameters.persist) {
-            // remove keyboard listener
-            this.cancelKeyboardResponse(listener_id);
+          if (!persist) {
+            // remove keyboard listener if it exists
+            this.cancelKeyboardResponse(listener);
           }
-        }
-      }
-    };
 
-    // create listener id object
-    listener_id = {
-      type: "keydown",
-      fn: listener_function,
+          callback_function({ key, rt });
+        }
+      },
     };
 
-    // add this keyboard listener to the list of listeners
-    this.keyboard_listeners.push(listener_id);
-
-    return listener_id;
+    this.listeners.add(listener);
+    return listener;
   }
 
-  cancelKeyboardResponse(listener) {
-    // remove the listener from the list of listeners
-    if (this.keyboard_listeners.includes(listener)) {
-      this.keyboard_listeners.splice(this.keyboard_listeners.indexOf(listener), 1);
-    }
+  cancelKeyboardResponse(listener: KeyboardListener) {
+    // remove the listener from the set of listeners if it is contained
+    this.listeners.delete(listener);
   }
 
   cancelAllKeyboardResponses() {
-    this.keyboard_listeners = [];
+    this.listeners.clear();
   }
 
-  convertKeyCharacterToKeyCode(character) {
+  convertKeyCharacterToKeyCode(character: string) {
```

Comment:
Leave the two methods or drop them in 7.0?

Reply:
I think we can drop them. There's no reference to them in any of the plugins at this point.